sdk/nodejs/test/connect.spec.ts

```diff
@@ -7,11 +7,27 @@ import * as os from "os"
 import * as path from "path"
 import * as tar from "tar"
 
+import * as dagger from "../api/client.gen.js"
 import { GraphQLRequestError } from "../common/errors/index.js"
-import { connect } from "../connect.js"
+import { connect, close } from "../connect.js"
 import * as bin from "../provisioning/bin.js"
 import { CLI_VERSION } from "../provisioning/default.js"
 
+describe("NodeJS default client", function () {
+  it("Should use the default client", async function () {
+    this.timeout(60000)
+
+    await dagger
+      .container()
+      .from("alpine")
+      .withExec(["apk", "add", "curl"])
+      .withExec(["curl", "https://dagger.io/"])
+      .sync()
+
+    close()
```

Comment:
Will users need to call `close()` manually like this?

Reply:
There's no defer logic in typescript so yeah, but we can probably workaround that with Zenith.
Isn't it the same in Python?

Reply:
## Python

In Python there's [multiple choices](https://github.com/dagger/dagger/blob/1f63a3cdd6fbe861c5637f965bc3cd96e24cd9d3/sdk/python/tests/client/test_default_client.py) but I want to remove some of the flexibility.

Preferred choice is to use a context manager (`with`), meaning that when code leaves the `with` block, it'll call `close` on whatever you want to manage (what you passed to `with`):
```python
with my_resource:
    ...
# at this point, my_resource is closed
```

### Lazy connection

For example, it's possible to use a lazy connection where the connection only gets established on first execute automatically. Then you have to remember to call `await dagger.close()` or wrap the calling code in a [context manager](https://dagger-io.readthedocs.io/en/sdk-python-v0.9.3/connection.html#dagger.closing).

The problem with this in practice is that it can become harder to debug when there's an issue. I hit several of these when I used this mode with python modules initially so I expect to remove this option.

You seem to only have `.close()` so are you doing a lazy connection? When does it get established?

### Explicit connection

It's much simpler to explicitly start the connection when you know you'll need it and let that manager close it in the end. That's how it's being used on python modules internally:

https://github.com/dagger/dagger/blob/1f63a3cdd6fbe861c5637f965bc3cd96e24cd9d3/sdk/python/src/dagger/mod/_module.py#L175-L176

In the example above, `dagger.connect()` doesn't provision. It only works in an existing session, like in Zenith or `dagger run`. It also doesn't accept `dagger.Config` because those settings are mostly about provisioning. 

### Legacy (provisioning) connection

There's an alternative connection function that does provisioning and accepts a `dagger.Config`:

```python
async with dagger.connection():
    ...
```

I added it for users running with `python main.py` (without `dagger run`), because otherwise they wouldn't be able to use the default client.

### New vs old

In any case, the big difference, if you ignore whether it does provisioning or not, is to not return a custom client that you need to pass around:

```diff
- async with dagger.Connection() as client:
+ async with dagger.connection():
-     await client.container()
+     await dagger.container()
``` 

## Node.JS

Which brings us to Node.JS. I suggest you keep `connect()` for non-Zenith (or rename it):
```js
import * as dagger from '@dagger.io/dagger'

dagger.connection(async () => {
    const out = await dagger
        .container()
        .from("alpine")
        .withExec(["echo"," hello", "world"])
        .stdout()

    console.log(out)
})
```

The advantage is:
- continue to support provisioning;
- explicit establishing of connection, but freedom do create reusable functions without `client`;
- closed automatically;

And since you can manage the connection behind the scenes on modules, then users only need to remove the wrapper to their function when they migrate (mostly):
```js
import * as dagger from '@dagger.io/dagger'

async function hello() {
    return await dagger
        .container()
        .from("alpine")
        .withExec(["echo"," hello", "world"])
        .stdout()
})
```

It's still much easier to gradually migrate to Zenith this way by first adopting the global client non-Zenith in a compatible way.

Reply:
> You seem to only have .close() so are you doing a lazy connection? When does it get established?

When the user is actually executing something so you are right, it's a lazy evalutation :D